In [1]:
from pathlib import Path
import sys

project_root = Path.cwd().parent

if str(project_root) not in sys.path:
    sys.path.append(str(project_root))

from src.data.load_data import (
    load_ai4i_dataset
)

from src.features.feature_engineering import (
    create_engineered_features
)

from src.features.preprocessing import (
    prepare_modeling_dataset
)

In [2]:
df = load_ai4i_dataset()

df = create_engineered_features(df)

df = prepare_modeling_dataset(df)

df.head()

,Air temperature [K],Process temperature [K],Rotational speed [rpm],Torque [Nm],Tool wear [min],Machine failure,Temperature Difference,Power Proxy,Mechanical Stress Index,Thermal Stress Index,Type_H,Type_L,Type_M
0,298.1,308.6,1551,42.8,0,0,10.5,66382.8,0.027595,449.40,0,0,1
1,298.2,308.7,1408,46.3,3,0,10.5,65190.4,0.032884,486.15,0,1,0
2,298.1,308.5,1498,49.4,5,0,10.4,74001.2,0.032977,513.76,0,1,0
3,298.2,308.6,1433,39.5,7,0,10.4,56603.5,0.027565,410.80,0,1,0
4,298.2,308.7,1408,40.0,9,0,10.5,56320.0,0.028409,420.00,0,1,0


In [3]:
df.columns.tolist()

['Air temperature [K]',
 'Process temperature [K]',
 'Rotational speed [rpm]',
 'Torque [Nm]',
 'Tool wear [min]',
 'Machine failure',
 'Temperature Difference',
 'Power Proxy',
 'Mechanical Stress Index',
 'Thermal Stress Index',
 'Type_H',
 'Type_L',
 'Type_M']

In [4]:
X = df.drop(
    columns=["Machine failure"]
)

y = df["Machine failure"]

In [5]:
print("Feature Matrix Shape:", X.shape)

print("Target Shape:", y.shape)

Feature Matrix Shape: (10000, 12)
Target Shape: (10000,)


In [6]:
y.value_counts()

Machine failure
0    9661
1     339
Name: count, dtype: int64

In [7]:
(
    y.value_counts(
        normalize=True
    ) * 100
).round(2)

Machine failure
0    96.61
1     3.39
Name: proportion, dtype: float64

In [8]:
X.head()

,Air temperature [K],Process temperature [K],Rotational speed [rpm],Torque [Nm],Tool wear [min],Temperature Difference,Power Proxy,Mechanical Stress Index,Thermal Stress Index,Type_H,Type_L,Type_M
0,298.1,308.6,1551,42.8,0,10.5,66382.8,0.027595,449.40,0,0,1
1,298.2,308.7,1408,46.3,3,10.5,65190.4,0.032884,486.15,0,1,0
2,298.1,308.5,1498,49.4,5,10.4,74001.2,0.032977,513.76,0,1,0
3,298.2,308.6,1433,39.5,7,10.4,56603.5,0.027565,410.80,0,1,0
4,298.2,308.7,1408,40.0,9,10.5,56320.0,0.028409,420.00,0,1,0


In [9]:
X.columns.tolist()

['Air temperature [K]',
 'Process temperature [K]',
 'Rotational speed [rpm]',
 'Torque [Nm]',
 'Tool wear [min]',
 'Temperature Difference',
 'Power Proxy',
 'Mechanical Stress Index',
 'Thermal Stress Index',
 'Type_H',
 'Type_L',
 'Type_M']

Modeling Dataset Summary

1. Identifier columns removed.

2. Failure-type columns removed to prevent target leakage.

3. Machine type encoded using one-hot encoding.

4. Engineered features retained.

5. Dataset prepared for model training.

6. Severe class imbalance (~3.4% failures) confirmed.